File: go/0238.product-of-array-except-self/solution.go
// Created by Olgierd Palasz at 2026/05/06 21:34
// leetgo: dev
// https://leetcode.com/problems/product-of-array-except-self/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func productExceptSelf(nums []int) []int {
	prefix := make([]int, len(nums))
	sufix := make([]int, len(nums))

	copy(prefix, nums)
	copy(sufix, nums)

	for i := 1; i < len(nums); i++ {
		prefix[i] *= prefix[i-1]
		sufix[len(nums)-1-i] *= sufix[len(nums)-i]
	}

	array := make([]int, len(nums))

	array[0] = sufix[1]
	array[len(array)-1] = prefix[len(array)-2]

	for i := 1; i < len(nums)-1; i++ {
		array[i] = prefix[i-1] * sufix[i+1]
	}
	return array
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	nums := Deserialize[[]int](ReadLine(stdin))
	ans := productExceptSelf(nums)

	fmt.Println("\noutput:", Serialize(ans))
}

File: rust/src/0238.product-of-array-except-self/solution.rs
// Created by Olgierd Palasz at 2026/07/16 08:04
// leetgo: dev
// https://leetcode.com/problems/product-of-array-except-self/

use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// @lc code=begin

impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let mut result: Vec<i32> = vec![1; nums.len()];

        let mut prefix_mul: i32 = 1;
        for i in 0..nums.len() {
            result[i] = prefix_mul;
            prefix_mul *= nums[i];
        }

        let mut suffix_mul: i32 = 1;
        for i in (0..nums.len()).rev() {
            result[i] *= suffix_mul;
            suffix_mul *= nums[i];
        }

        result
    }
}

// @lc code=end

fn main() -> Result<()> {
    let nums: Vec<i32> = deserialize(&read_line()?)?;
    let ans: Vec<i32> = Solution::product_except_self(nums);

    println!("\noutput: {}", serialize(ans)?);
    Ok(())
}
